File: JumpGameIIV2.cpp
#include <vector>
#include <cstring>
#include <iostream>
using namespace std;
// AC
// bfs
class Solution {
public:
    int jump(vector<int>& nums) {
        if(nums.size()<=1)return 0;
        int i=0,curMax=0,step=0,nextMax=0;
        while(curMax>=i){
            for(;i<=curMax;i++){
                nextMax=max(nextMax,i+nums[i]);
                if(nextMax>=nums.size()-1){
                    return step+1;
                }
            }
            step++;
            curMax=nextMax;
        }
        return -1;
    }
};

int main(){
    Solution sol;
    vector<int>v;
    int temp=1;
    // freopen("./JumpGameII.in","r",stdin);
    freopen("./test.in","r",stdin);
    while(scanf("%d",&temp)==1){
        v.push_back(temp);
    }
    cout<<sol.jump(v)<<endl;

    return 0;
}

File: Jump_Game_II_v2.kt
import java.lang.Integer.max

class Solution {
    fun jump(nums: IntArray): Int {
        if (nums.isEmpty() || nums.size == 1) return 0
        var step = 0
        var cur = 0
        var longest = 0
        while (longest < nums.lastIndex) {
            step++
            val temp = longest
            for (i in cur..longest) {
                longest = max(longest, i + nums[i])
                if (longest >= nums.lastIndex) break
            }
            cur = temp
        }
        return step
    }
}

fun main() {
    val solution = Solution()
    // 2 2 0 2
    val testCases = arrayOf(intArrayOf(2, 3, 1, 1, 4), intArrayOf(2, 3, 0, 1, 4), intArrayOf(1), intArrayOf(7, 0, 9, 6, 9, 6, 1, 7, 9, 0, 1, 2, 9, 0, 3))
    for (case in testCases) {
        println(solution.jump(case))
    }
}
